data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1728 @@
+
+    delete rbt;
+    cout << "PASSED!" << endl << endl;
+}
+
+/**
+ * @brief testing removing a value with a "red sibling"
+ * from tree
+ * @returns void
+ */
+void RemoveTestRedSiblingCase() {
+    cout << "Testing remove that requires double-black fix with red "
+            "sibling case..."
+         << endl;
+
+    RedBlackTree rbt = RedBlackTree();
+
+    int fill_tree_vec[6] = {11, 9, 7, 5, 8, 6};
+    for (int val : fill_tree_vec) {
+        rbt.Insert(val);
+    }
+    cout << "tree: " << rbt.ToPrefixString() << endl;
+    assert(rbt.ToPrefixString() == " B9  R7  B5  R6  B8  B11 ");
+
+    rbt.Remove(11);  // 11 is black leaf with red sibling 7
+    cout << "removing 11: " << rbt.ToPrefixString() << endl;
+    assert(rbt.ToPrefixString() == " B7  B5  R6  B9  R8 ");
+
+    cout << "PASSED!" << endl << endl;
+}
+
+/**
+ * @brief testing memory management as values are
+ * removed from tree
+ * @returns void
+ */
+void LargeRandomRemoveMemoryTest() {
+    cout << "Giant Remove Test..." << endl;
+
+    RedBlackTree* giant_rbt = new RedBlackTree();
+
+    vector<int> random_numbers;  // to store inserted numbers
+    unsigned int seed = time(NULL);
+
+    while (giant_rbt->Size() != 711) {
+        int random_num = rand_r(&seed);
+        if (!giant_rbt->Contains(random_num)) {
+            random_numbers.push_back(random_num);
+            giant_rbt->Insert(random_num);
+        }
+    }
+    // getting numbers to remove, starting from end of vector
+    size_t idx = random_numbers.size() - 1;
+
+    // then removing the numbers until tree is empty
+    // (checking for memory issues/exceptions)
+    while (giant_rbt->Size() != 0) {
+        giant_rbt->Remove(random_numbers[idx--]);
+    }
+
+    delete giant_rbt;
+    cout << "PASSED!" << endl << endl;
+}
+
+/**
+ * @brief more testing of removing certain nodes from
+ * a tree
+ * @returns void
+ */
+void LargerComprehensiveRemoveTest() {
+    cout << "Testing many removes from larger tree..." << endl;
+
+    RedBlackTree* rbt = new RedBlackTree();
+
+    int fill_tree_vec[19] = {40, 97, 12, 43,   33,   87,   99,  439, 8762, 3,
+                             7,  1,  2,  9999, 8888, 2345, 111, 687, 0};
+
+    for (int val : fill_tree_vec) {
+        rbt->Insert(val);
+    }
+
+    cout << "tree: " << rbt->ToPrefixString() << endl;
+    assert(rbt->ToPrefixString() ==
+           " B87  B12  R3  B1  R0  R2  B7  R40  B33"
+           "  B43  B439  R99  B97  B111  R8762  B2345  R687  B9999  R8888 ");
+
+    rbt->Remove(87);  // removing root
+    cout << "removing 87: " << rbt->ToPrefixString() << endl;
+    assert(rbt->ToPrefixString() ==
+           " B97  B12  R3  B1  R0  R2  B7  R40  B33"
+           "  B43  B439  B99  R111  R8762  B2345  R687  B9999  R8888 ");
+
+    delete rbt;
+    cout << "PASSED!" << endl << endl;
+}
+
+/**
+ * @brief main function where all test functions are
+ * called
+ * @returns void
+ */
+int main() {
+    TestSimpleConstructor();
```

Comment:
i think moving all of these tests into a function with calls all of them would be nice
tests(); // runs self test implementations